File: scripts/prepare_site.py
#!/usr/bin/env python3
"""Copy markdown content and generate navigation manifest for the static site."""

import json
import re
import shutil
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SITE = ROOT / "site"
CONTENT_OUT = SITE / "public" / "content"
NAV_OUT = SITE / "src" / "data" / "navigation.json"

WEEK_TOPICS = {
    1: "C# Language Mastery",
    2: ".NET Runtime & Ecosystem",
    3: "SOLID & Clean Architecture",
    4: "Design Patterns",
    5: "Data Structures",
    6: "Algorithms & Complexity",
    7: "SQL Server",
    8: "PostgreSQL",
    9: "Azure Fundamentals",
    10: "Azure Compute",
    11: "Azure Data Platform",
    12: "Azure Identity",
    13: "Azure Networking",
    14: "Azure Security",
    15: "Azure Integration",
    16: "Azure Capstone",
    17: "AWS Fundamentals",
    18: "AWS Compute",
    19: "AWS Data & VPC",
    20: "Multi-Cloud",
    21: "Distributed Systems",
    22: "Microservices",
    23: "Domain-Driven Design",
    24: "Microservices Capstone",
    25: "Docker & Containers",
    26: "Kubernetes Fundamentals",
    27: "Kubernetes Advanced",
    28: "Linux for Architects",
    29: "DevOps Culture",
    30: "CI/CD Pipelines",
    31: "Infrastructure as Code",
    32: "Observability",
    33: "System Design",
    34: "Scalability & Caching",
    35: "Messaging & Events",
    36: "System Design Capstone",
    37: "AI/ML Architecture",
    38: "RAG & Vector Search",
    39: "LLMOps",
    40: "AI Capstone",
    41: "Architecture Reviews",
    42: "FinOps & DR",
    43: "Frontend Architecture",
    44: "Enterprise Case Studies",
    45: "Interview Intensive I",
    46: "Interview Intensive II",
    47: "Interview Intensive III",
    48: "Whiteboard Mocks I",
    49: "Whiteboard Mocks II",
    50: "Leadership I",
    51: "Leadership II",
    52: "Graduation Capstone",
}

PHASE_NAMES = {
    1: "Month 1 — .NET Foundation",
    2: "Month 2 — Data & Algorithms",
    3: "Month 3 — Azure Core",
    4: "Month 4 — Azure Deep Dive",
    5: "Month 5 — AWS & Multi-Cloud",
    6: "Month 6 — Distributed Systems",
    7: "Month 7 — Containers & K8s",
    8: "Month 8 — DevOps",
    9: "Month 9 — System Design",
    10: "Month 10 — AI Architecture",
    11: "Month 11 — Enterprise",
    12: "Month 12 — Interview & Graduation",
}

WEEK_SECTIONS = [
    ("README.md", "Overview"),
    ("theory", "Theory"),
    ("diagrams", "Diagrams"),
    ("labs", "Labs"),
    ("exercises", "Exercises"),
    ("interview-questions", "Interview Q&A"),
    ("case-studies", "Case Studies"),
    ("assessments", "Assessments"),
]


def doc_path(rel: str) -> str:
    rel = rel.replace("\\", "/")
    if rel.endswith(".md"):
        rel = rel[:-3]
    if rel.endswith("/README"):
        rel = rel[: -len("/README")]
    return f"/doc/{rel}" if rel else "/doc"


def item(title: str, path: str, children=None):
    entry = {"title": title, "path": path}
    if children:
        entry["children"] = children
    return entry


def week_children(week_num: int) -> list:
    week_dir = ROOT / f"weeks/week-{week_num:02d}"
    if not week_dir.exists():
        return []
    base = f"weeks/week-{week_num:02d}"
    children = []
    for segment, label in WEEK_SECTIONS:
        if segment.endswith(".md"):
            if segment == "README.md":
                continue  # week overview = parent nav item
            p = week_dir / segment
            if p.exists():
                children.append(item(label, doc_path(f"{base}/{segment[:-3]}")))
        else:
            folder = week_dir / segment
            if folder.exists():
                sub = []
                section_path = doc_path(f"{base}/{segment}/README")
                for md in sorted(folder.glob("*.md")):
                    if md.name == "README.md":
                        continue
                    name = md.stem.replace("-", " ").title()
                    sub.append(item(name, doc_path(f"{base}/{segment}/{md.stem}")))
                if sub or (folder / "README.md").exists():
                    children.append(item(label, section_path, sub if sub else None))
    cm = week_dir / "common-mistakes.md"
    if cm.exists():
        children.append(item("Common Mistakes", doc_path(f"{base}/common-mistakes")))
    return children


def scan_dir_nav(folder: Path, base_path: str, max_depth: int = 2) -> list:
    if not folder.exists():
        return []
    items = []
    if max_depth <= 0:
        return items
    for child in sorted(folder.iterdir()):
        if child.is_dir() and not child.name.startswith("."):
            sub_path = f"{base_path}/{child.name}" if base_path else child.name
            readme = child / "README.md"
            md_files = [f for f in sorted(child.glob("*.md")) if f.name != "README.md"]
            sub_items = scan_dir_nav(child, sub_path, max_depth - 1) if max_depth > 1 else []

            if readme.exists():
                items.append(
                    item(
                        child.name.replace("-", " ").title(),
                        doc_path(f"{sub_path}/README"),
                        sub_items or None,
                    )
                )
            elif md_files:
                file_children = [
                    item(f.stem.replace("-", " ").title(), doc_path(f"{sub_path}/{f.stem}"))
                    for f in md_files
                ]
                items.append(
                    item(
                        child.name.replace("-", " ").title(),
                        file_children[0]["path"],
                        file_children if len(file_children) > 1 else None,
                    )
                )
            elif sub_items:
                items.append(
                    item(
                        child.name.replace("-", " ").title(),
                        sub_items[0]["path"],
                        sub_items if len(sub_items) > 1 else None,
                    )
                )
        elif child.suffix == ".md" and child.name != "README.md":
            rel = f"{base_path}/{child.stem}" if base_path else child.stem
            items.append(item(child.stem.replace("-", " ").title(), doc_path(rel)))
    return items


def build_navigation() -> dict:
    sections = []

    sections.append({
        "title": "Start",
        "items": [
            item("Home", "/"),
            item("How to Use", doc_path("docs/getting-started/how-to-use-this-guide")),
            item("Syllabus", doc_path("SYLLABUS")),
            item("Curriculum Map", doc_path("CURRICULUM-MAP")),
            item("12-Month Study Plan", doc_path("STUDY-PLAN-12-MONTHS")),
        ],
    })

    # Weeks by phase
    for phase in range(1, 13):
        week_items = []
        for w in range((phase - 1) * 4 + 1, min(phase * 4 + 1, 53)):
            if w > 52:
                break
            topic = WEEK_TOPICS.get(w, f"Week {w}")
            week_items.append(
                item(
                    f"Week {w:02d} — {topic}",
                    doc_path(f"weeks/week-{w:02d}/README"),
                    week_children(w),
                )
            )
        sections.append({
            "title": PHASE_NAMES[phase],
            "items": week_items,
        })

    sections.append({
        "title": "Program",
        "items": [
            item(f"Phase {i} Capstone", doc_path(f"program/phase-{i:02d}-month-{i:02d}/capstone"))
            for i in range(1, 13)
        ],
    })

    sections.append({
        "title": "Modules",
        "items": scan_dir_nav(ROOT / "modules", "modules", max_depth=1),
    })

    sections.append({
        "title": "Interview Prep",
        "items": scan_dir_nav(ROOT / "interview-prep", "interview-prep", max_depth=1),
    })

    sections.append({
        "title": "Documentation",
        "items": scan_dir_nav(ROOT / "docs", "docs", max_depth=2),
    })

    sections.append({
        "title": "Resources",
        "items": [
            item("Case Studies Index", doc_path("case-studies/README")),
            item("Templates", doc_path("templates/case-study-template")),
        ],
    })

    return {"sections": sections}


def copy_markdown():
    if CONTENT_OUT.exists():
        shutil.rmtree(CONTENT_OUT)
    CONTENT_OUT.mkdir(parents=True, exist_ok=True)

    skip_dirs = {".git", "site", "node_modules", ".cursor", "scripts"}
    count = 0
    for md in ROOT.rglob("*.md"):
        parts = set(md.relative_to(ROOT).parts)
        if parts & skip_dirs:
            continue
        rel = md.relative_to(ROOT)
        dest = CONTENT_OUT / rel
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(md, dest)
        count += 1
    return count


def main():
    NAV_OUT.parent.mkdir(parents=True, exist_ok=True)
    nav = build_navigation()
    NAV_OUT.write_text(json.dumps(nav, indent=2), encoding="utf-8")
    count = copy_markdown()
    print(f"Navigation written to {NAV_OUT.relative_to(ROOT)}")
    print(f"Copied {count} markdown files to {CONTENT_OUT.relative_to(ROOT)}")


if __name__ == "__main__":
    main()

File: site/scripts/prepare-site.mjs
#!/usr/bin/env node
/**
 * Copy markdown content and generate navigation.json for the static site.
 * Run from site/ — curriculum root is one level up (full repo required on Vercel).
 */
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const SITE = path.resolve(__dirname, '..')
const ROOT = path.resolve(SITE, '..')
const CONTENT_OUT = path.join(SITE, 'public', 'content')
const NAV_OUT = path.join(SITE, 'src', 'data', 'navigation.json')

const WEEK_TOPICS = {
  1: 'C# Language Mastery', 2: '.NET Runtime & Ecosystem', 3: 'SOLID & Clean Architecture',
  4: 'Design Patterns', 5: 'Data Structures', 6: 'Algorithms & Complexity', 7: 'SQL Server',
  8: 'PostgreSQL', 9: 'Azure Fundamentals', 10: 'Azure Compute', 11: 'Azure Data Platform',
  12: 'Azure Identity', 13: 'Azure Networking', 14: 'Azure Security', 15: 'Azure Integration',
  16: 'Azure Capstone', 17: 'AWS Fundamentals', 18: 'AWS Compute', 19: 'AWS Data & VPC',
  20: 'Multi-Cloud', 21: 'Distributed Systems', 22: 'Microservices', 23: 'Domain-Driven Design',
  24: 'Microservices Capstone', 25: 'Docker & Containers', 26: 'Kubernetes Fundamentals',
  27: 'Kubernetes Advanced', 28: 'Linux for Architects', 29: 'DevOps Culture', 30: 'CI/CD Pipelines',
  31: 'Infrastructure as Code', 32: 'Observability', 33: 'System Design', 34: 'Scalability & Caching',
  35: 'Messaging & Events', 36: 'System Design Capstone', 37: 'AI/ML Architecture',
  38: 'RAG & Vector Search', 39: 'LLMOps', 40: 'AI Capstone', 41: 'Architecture Reviews',
  42: 'FinOps & DR', 43: 'Frontend Architecture', 44: 'Enterprise Case Studies',
  45: 'Interview Intensive I', 46: 'Interview Intensive II', 47: 'Interview Intensive III',
  48: 'Whiteboard Mocks I', 49: 'Whiteboard Mocks II', 50: 'Leadership I', 51: 'Leadership II',
  52: 'Graduation Capstone',
}

const PHASE_NAMES = {
  1: 'Month 1 — .NET Foundation', 2: 'Month 2 — Data & Algorithms', 3: 'Month 3 — Azure Core',
  4: 'Month 4 — Azure Deep Dive', 5: 'Month 5 — AWS & Multi-Cloud', 6: 'Month 6 — Distributed Systems',
  7: 'Month 7 — Containers & K8s', 8: 'Month 8 — DevOps', 9: 'Month 9 — System Design',
  10: 'Month 10 — AI Architecture', 11: 'Month 11 — Enterprise', 12: 'Month 12 — Interview & Graduation',
}

const WEEK_SECTIONS = [
  ['README.md', 'Overview'], ['theory', 'Theory'], ['diagrams', 'Diagrams'], ['labs', 'Labs'],
  ['exercises', 'Exercises'], ['interview-questions', 'Interview Q&A'], ['case-studies', 'Case Studies'],
  ['assessments', 'Assessments'],
]

const SKIP_DIRS = new Set(['.git', 'site', 'node_modules', '.cursor', 'scripts'])

function docPath(rel) {
  rel = rel.replace(/\\/g, '/')
  if (rel.endsWith('.md')) rel = rel.slice(0, -3)
  if (rel.endsWith('/README')) rel = rel.slice(0, -'/README'.length)
  return rel ? `/doc/${rel}` : '/doc'
}

function navItem(title, p, children) {
  const entry = { title, path: p }
  if (children?.length) entry.children = children
  return entry
}

function weekChildren(weekNum) {
  const weekDir = path.join(ROOT, `weeks/week-${String(weekNum).padStart(2, '0')}`)
  if (!fs.existsSync(weekDir)) return []
  const base = `weeks/week-${String(weekNum).padStart(2, '0')}`
  const children = []
  for (const [segment, label] of WEEK_SECTIONS) {
    if (segment.endsWith('.md')) {
      if (segment === 'README.md') continue
      if (fs.existsSync(path.join(weekDir, segment))) {
        children.push(navItem(label, docPath(`${base}/${segment.slice(0, -3)}`)))
      }
    } else {
      const folder = path.join(weekDir, segment)
      if (!fs.existsSync(folder)) continue
      const sub = []
      const sectionPath = docPath(`${base}/${segment}/README`)
      for (const name of fs.readdirSync(folder).filter((f) => f.endsWith('.md') && f !== 'README.md').sort()) {
        const stem = name.replace(/\.md$/, '')
        sub.push(navItem(stem.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()), docPath(`${base}/${segment}/${stem}`)))
      }
      if (sub.length || fs.existsSync(path.join(folder, 'README.md'))) {
        children.push(navItem(label, sectionPath, sub.length ? sub : undefined))
      }
    }
  }
  if (fs.existsSync(path.join(weekDir, 'common-mistakes.md'))) {
    children.push(navItem('Common Mistakes', docPath(`${base}/common-mistakes`)))
  }
  return children
}

function scanDirNav(folder, basePath, maxDepth = 2) {
  if (!fs.existsSync(folder) || maxDepth <= 0) return []
  const items = []
  for (const name of fs.readdirSync(folder).sort()) {
    const child = path.join(folder, name)
    const stat = fs.statSync(child)
    if (stat.isDirectory() && !name.startsWith('.')) {
      const subPath = basePath ? `${basePath}/${name}` : name
      const readme = path.join(child, 'README.md')
      const mdFiles = fs.readdirSync(child).filter((f) => f.endsWith('.md') && f !== 'README.md').sort()
      const subItems = maxDepth > 1 ? scanDirNav(child, subPath, maxDepth - 1) : []
      if (fs.existsSync(readme)) {
        items.push(navItem(name.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()), docPath(`${subPath}/README`), subItems.length ? subItems : undefined))
      } else if (mdFiles.length) {
        const fileChildren = mdFiles.map((f) => {
          const stem = f.replace(/\.md$/, '')
          return navItem(stem.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()), docPath(`${subPath}/${stem}`))
        })
        items.push(navItem(name.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()), fileChildren[0].path, fileChildren.length > 1 ? fileChildren : undefined))
      } else if (subItems.length) {
        items.push(navItem(name.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()), subItems[0].path, subItems.length > 1 ? subItems : undefined))
      }
    } else if (name.endsWith('.md') && name !== 'README.md') {
      const stem = name.replace(/\.md$/, '')
      const rel = basePath ? `${basePath}/${stem}` : stem
      items.push(navItem(stem.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()), docPath(rel)))
    }
  }
  return items
}

function buildNavigation() {
  const sections = [{
    title: 'Start',
    items: [
      navItem('Home', '/'),
      navItem('How to Use', docPath('docs/getting-started/how-to-use-this-guide')),
      navItem('Syllabus', docPath('SYLLABUS')),
      navItem('Curriculum Map', docPath('CURRICULUM-MAP')),
      navItem('12-Month Study Plan', docPath('STUDY-PLAN-12-MONTHS')),
    ],
  }]
  for (let phase = 1; phase <= 12; phase++) {
    const weekItems = []
    for (let w = (phase - 1) * 4 + 1; w <= phase * 4 && w <= 52; w++) {
      weekItems.push(navItem(`Week ${String(w).padStart(2, '0')} — ${WEEK_TOPICS[w] || `Week ${w}`}`, docPath(`weeks/week-${String(w).padStart(2, '0')}/README`), weekChildren(w)))
    }
    sections.push({ title: PHASE_NAMES[phase], items: weekItems })
  }
  sections.push({
    title: 'Program',
    items: Array.from({ length: 12 }, (_, i) => navItem(`Phase ${i + 1} Capstone`, docPath(`program/phase-${String(i + 1).padStart(2, '0')}-month-${String(i + 1).padStart(2, '0')}/capstone`))),
  })
  sections.push({ title: 'Modules', items: scanDirNav(path.join(ROOT, 'modules'), 'modules', 1) })
  sections.push({ title: 'Interview Prep', items: scanDirNav(path.join(ROOT, 'interview-prep'), 'interview-prep', 1) })
  sections.push({ title: 'Documentation', items: scanDirNav(path.join(ROOT, 'docs'), 'docs', 2) })
  sections.push({
    title: 'Resources',
    items: [navItem('Case Studies Index', docPath('case-studies/README')), navItem('Templates', docPath('templates/case-study-template'))],
  })
  return { sections }
}

function copyMarkdown() {
  if (fs.existsSync(CONTENT_OUT)) fs.rmSync(CONTENT_OUT, { recursive: true })
  fs.mkdirSync(CONTENT_OUT, { recursive: true })
  let count = 0
  function walk(dir) {
    for (const name of fs.readdirSync(dir)) {
      const full = path.join(dir, name)
      const relParts = path.relative(ROOT, full).split(path.sep)
      if (relParts.some((p) => SKIP_DIRS.has(p))) continue
      const stat = fs.statSync(full)
      if (stat.isDirectory()) walk(full)
      else if (name.endsWith('.md')) {
        const rel = path.relative(ROOT, full)
        const dest = path.join(CONTENT_OUT, rel)
        fs.mkdirSync(path.dirname(dest), { recursive: true })
        fs.copyFileSync(full, dest)
        count++
      }
    }
  }
  walk(ROOT)
  return count
}

function main() {
  if (!fs.existsSync(path.join(ROOT, 'weeks'))) {
    console.error('')
    console.error('ERROR: Curriculum folder not found at', path.join(ROOT, 'weeks'))
    console.error('')
    console.error('Vercel must deploy the FULL repo with Root Directory = site.')
    console.error('Run:  cd ArchitectGuide   (repo root, not site/)')
    console.error('      vercel link')
    console.error('      vercel --prod')
    console.error('')
    process.exit(1)
  }
  fs.mkdirSync(path.dirname(NAV_OUT), { recursive: true })
  fs.writeFileSync(NAV_OUT, JSON.stringify(buildNavigation(), null, 2))
  const count = copyMarkdown()
  console.log(`Navigation written to src/data/navigation.json`)
  console.log(`Copied ${count} markdown files to public/content`)
}

main()
